chainerx_cc/chainerx/CMakeLists.txt

```diff
@@ -84,20 +84,38 @@ add_library(chainerx SHARED
     thread_local_state.cc
     )
 
-target_link_libraries(chainerx
-    PUBLIC
-    ${CMAKE_DL_LIBS}
-    PRIVATE
-    -Wl,--whole-archive)
-target_link_libraries(chainerx PRIVATE chainerx_routines)
-target_link_libraries(chainerx PRIVATE chainerx_native)
-target_link_libraries(chainerx PRIVATE chainerx_testing)
-if(${CUDA_FOUND})
-    target_link_libraries(chainerx PRIVATE chainerx_cuda)
+if(${APPLE})
+  target_link_libraries(chainerx
+      PUBLIC
+      ${CMAKE_DL_LIBS})
+
+  target_link_libraries(chainerx PRIVATE -Wl,-force_load chainerx_routines)
+  target_link_libraries(chainerx PRIVATE -Wl,-force_load chainerx_native)
+  target_link_libraries(chainerx PRIVATE -Wl,-force_load chainerx_testing)
+
+  if(${CUDA_FOUND})
+    target_link_libraries(chainerx PRIVATE -Wl,-force_load chainerx_cuda)
+  endif()
+
+  target_link_libraries(chainerx PRIVATE -Wl,-noall_load)
+else()
+  target_link_libraries(chainerx
+      PUBLIC
+      ${CMAKE_DL_LIBS}
+      PRIVATE
+      -Wl,--whole-archive)
+
+  target_link_libraries(chainerx PRIVATE chainerx_routines)
+  target_link_libraries(chainerx PRIVATE chainerx_native)
+  target_link_libraries(chainerx PRIVATE chainerx_testing)
+
+  if(${CUDA_FOUND})
+      target_link_libraries(chainerx PRIVATE chainerx_cuda)
+  endif()
+
+  target_link_libraries(chainerx PRIVATE -Wl,--no-whole-archive)
```

Comment:
What do you think of unifying the `target_link_libraries` as follows, for instance.
```cmake
if(${APPLE})
  target_link_libraries(chainerx
      PUBLIC
      ${CMAKE_DL_LIBS}
      PRIVATE
      -Wl,-all_load)
else()
  target_link_libraries(chainerx
      PUBLIC
      ${CMAKE_DL_LIBS}
      PRIVATE
      -Wl,--whole-archive)
endif()

target_link_libraries(chainerx PRIVATE chainerx_routines)
target_link_libraries(chainerx PRIVATE chainerx_native)
target_link_libraries(chainerx PRIVATE chainerx_testing)

if(${CUDA_FOUND})
    target_link_libraries(chainerx PRIVATE chainerx_cuda)
endif()

if(${APPLE})
  target_link_libraries(chainerx PRIVATE -Wl,-noall_load)
else()
  target_link_libraries(chainerx PRIVATE -Wl,--no-whole-archive)
endif()
```

Reply:
A nitpick but could you align the indentation? It might be the case that it's not correct everywhere else but we are trying to stick to 4 spaces.

Reply:
FWIU the `-all_load` flag will load all members from all static libraries linked to the binary and not just the ones that follow the flag. Hence the cluttering. I'm not too sure if my understanding is correct though.

From the ld manual:

```
     -all_load   Loads all members of static archive libraries.

     -ObjC       Loads all members of static archive libraries that implement an Objective-C class or category.

     -force_load path_to_archive
                 Loads all members of the specified static archive library.  Note: -all_load forces all members of all archives to be loaded.  This option allows
                 you to target a specific archive.
```

Because the documentation for `-force_load` explicitly states that it is for designating specific archives I assumed that `-all_load` will probably behave in an unwanted way.

Reply:
Thanks for the reference. I checked the sizes which seemed to be identical in this case, but I see that `-force_load` would be preferred. Let's keep it as is.